### PR TITLE
SlurmGCP. Perform cleanup per nodeset

### PR DESCRIPTION
**Pull request by mr0re1**
* Do clean up per nodeset instead of total:
  * solves problem "deletion of nodeset with networks causes total wipe";
  * solves problem "dynamic nodes getting deleted";
  * pre-requisite for reversing chain of dependencies between modules;
* Remove controller dependency on clean up - solves #2774 ;
* Do not participate in "race against controller":
  * Fetch list on instances to delete once;
  * Do not fail on failed attempt to delete;
  * Wait until `STOPPING` instances are deleted.

**Remaining problems:**
* Turning `enable_cleanup_compute`  off triggers clean up;
* "Slow to die" controller or removal of one partition only have an opportunity to leak some instances (or fail on policies deletion).

### CODE REVIEW

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/cleanup_compute.sh [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+set -e -o pipefail
+
+project="$1"
+cluster_name="$2"
+nodeset_name="$3"
+universe_domain="$4"
+compute_endpoint_version="$5"
+gcloud_dir="$6"
+
+if [[ -z "${project}" || -z "${cluster_name}" || -z "${nodeset_name}" || -z "${universe_domain}" || -z "${compute_endpoint_version}" ]]; then
```
Comment (tpdownes) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/cleanup_compute.sh:25]: `nit` you might add an argument count check here instead. e.g.

https://stackoverflow.com/a/18568726

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/cleanup_compute.sh [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+gcloud compute instances list --format="value(selfLink)" --filter="${running_nodes_filter}" >"$tmpfile"
+# Do 10 instances at a time
+while batch="$(head -n 10)" && [[ ${#batch} -gt 0 ]]; do
+	nodes=$(echo "$batch" | paste -sd " " -) # concat into a single space-separated line
+	# The lack of quotes around ${nodes} is intentional and causes each new space-separated "word" to
+	# be treated as independent arguments. See PR#2523
+	# shellcheck disable=SC2086
+	gcloud compute instances delete --quiet ${nodes} || echo "Failed to delete some instances"
+done <"$tmpfile" # have to use a temp file, since `< <(gcloud ...)` doesn't work nicely with `head`
+rm -f "$tmpfile"
```
Comment (tpdownes) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/cleanup_compute.sh:58]: I suggest a `trap` for this immediately after creating the file. e.g.

`trap 'rm -rf -- "$tmpfile"' EXIT`

(confirm this first, I'm doing a bit of cut-and-paste from example)

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/variables.tf [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+Enables automatic cleanup of compute nodes and resource policies (e.g.
+placement groups) managed by this module, when cluster is destroyed.
+
+*WARNING*: Toggling this off will impact the running workload.
+Deployed compute nodes will be destroyed.
+EOD
+  type        = bool
+}
+
+variable "universe_domain" {
```
Comment (tpdownes) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/variables.tf:39]: It seems like we either need default values for these API configurations here or in the bash script. Am I not seeing where the default/simple case (GA API, standard universe) is being set?

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/main.tf [hunk 4]
```diff
@@ -14,36 +14,32 @@
...
+    cluster_name             = var.slurm_cluster_name
+    nodeset_name             = var.nodeset.nodeset_name
     universe_domain          = var.universe_domain
     compute_endpoint_version = var.endpoint_versions.compute
     gcloud_path_override     = var.gcloud_path_override
   }
 
   provisioner "local-exec" {
-    command = "/bin/bash ${path.module}/scripts/cleanup_compute.sh ${self.triggers.project_id} ${self.triggers.cluster_name} ${self.triggers.universe_domain} ${self.triggers.compute_endpoint_version} ${self.triggers.gcloud_path_override}"
+    command = "/bin/bash ${path.module}/cleanup_compute.sh ${self.triggers.project_id} ${self.triggers.cluster_name} ${self.triggers.nodeset_name} ${self.triggers.universe_domain} ${self.triggers.compute_endpoint_version} ${self.triggers.gcloud_path_override}"
```
Comment (tpdownes) [inline community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/main.tf:39]: I'd like to consider having this be in a subdirectory as it was before. Let me review best practices for storage of local-exec scripts.

**Review by tpdownes — CHANGES_REQUESTED** [review]
- There are some minor bash script changes I'd like to see.
- A couple v6 integration tests were skipped, can you please run them?

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/main.tf (hunk 4 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/main.tf]: Moved to `scripts/`

File: community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/variables.tf (hunk 3 above)
Comment (mr0re1) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/variables.tf:39]: It's in `community/modules/scheduler/schedmd-slurm-gcp-v6-controller/variables.tf`
I would prefer to have at most one place were defaults are defined, and touching controller module interface increases scope needlessly.

Comment (tpdownes) [reply community/modules/scheduler/schedmd-slurm-gcp-v6-controller/modules/cleanup_compute/variables.tf:39]: I agree, I just couldn't find it via GitHub UI easily. I also think this is a case where the sub-module is an internal implementation detail so exposing the default at the top-level is the right thing.

**Review by tpdownes — APPROVED** [review]

**Pull request merged**